packages/jest-matchers/src/matchers.js

```diff
@@ -408,6 +408,76 @@ const matchers: MatchersObject = {
   },
 
   toMatchSnapshot,
+
+  toMatchObject(receivedObject: Object, expectedObject: Object) {
+
+    if (typeof receivedObject !== 'object' || receivedObject === null) {
+      throw new Error(
+        matcherHint('[.not].toMatch', 'object', 'expected') + '\n\n' +
+        `${RECEIVED_COLOR('string')} value must be an object.\n` +
+        printWithType('Received', JSON.stringify(receivedObject), printReceived),
+      );
+    }
+
+    if (typeof expectedObject !== 'object' || expectedObject === null) {
+      throw new Error(
+        matcherHint('[.not].toMatch', 'object', 'expected') + '\n\n' +
+        `${RECEIVED_COLOR('string')} value must be an object.\n` +
+        printWithType('Received', JSON.stringify(expectedObject), printReceived),
+      );
+    }
+
+    const compare = (expected: any, actual: any) => {
+
+      if (typeof actual !== typeof expected) {
+        return false;
+      }
+      if (typeof expected !== 'object' || expected === null) {
+        return expected === actual;
+      }
+
+      if (Array.isArray(expected)) {
+        if (!Array.isArray(actual)) {
+          return false;
+        }
+
+        return expected.every(exp => {
+          return actual.some(act => {
+            return compare(exp, act);
+          });
+        });
+      }
+
+      if (expected instanceof Date && actual instanceof Date) {
+        return expected.getTime() === actual.getTime();
+      }
+
+      return Object.keys(expected).every(key => {
+        const exp = expected[key];
+        const act = actual[key];
+        if (typeof exp === 'object' && exp !== null && act !== null) {
+          return compare(exp, act);
+        }
+        return act === exp;
+      });
+    };
+
+    const pass = compare(expectedObject, receivedObject);
+
+    const message = pass
+      ? () => matcherHint('.not.toMatchObject') +
+     `\n\nExpected value not to match object:\n` +
+     `  ${printExpected(JSON.stringify(expectedObject))}` +
```

Comment:
we can't really use `JSON.stringify` here. 
you should use `printExpected` and `printReceived`, they won't break on circular references and will handle other special values like `undefined` and friends :)

i think what we should also print is a diff between the entire `expected` and filtered `received` object, with only properties that matter. something like:

```js
expect({a: 1, b: 1, c: 1, d: {e: {f: 555}}}).toMatch({d: {e: {f: 222}}});

Expected object to match:
  {d: {e: {f: 222}}}
Received:
  {a: 1, b: 1, c: 1, d: {e: {f: 555}}}
Difference:
  {
     d: {
       e: {
+         f:  555,
-          f: 222
        }
     }
  }
```